### PR TITLE
`k6 cloud` command

### LINKED ISSUE #366: Cloud execution integration
Issue to track work on the integration between k6 and Load Impact's cloud service to run distributed load tests. This work is being done in the https://github.com/loadimpact/k6/tree/feature/cloud-exec branch.

### PR DESCRIPTION
**Pull request by ppcano**
Implements the command to execute a `k6` script onto the Load Impact cloud service.

`k6 cloud script.js`

Closes #366

### CODE REVIEW

File: cmd/cloud.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/loadimpact/k6/stats/cloud"
+	"github.com/loadimpact/k6/ui"
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+var cloudCmd = &cobra.Command{
+	Use:   "cloud",
+	Short: "Execute a test on the cloud",
```
Comment (robingustafsson) [inline cmd/cloud.go]: Maybe "Execute a test in the cloud on infrastructure managed by Load Impact"?

File: cmd/cloud.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	Long: `Execute a test on the cloud.
+
+	This will execute the test on the Load Impact cloud service. Use "k6 login cloud" to authenticate.`,
```
Comment (robingustafsson) [inline cmd/cloud.go]: Maybe "Execute a test on infrastructure managed by the Load Impact cloud service. Use "k6 login cloud" to authenticate."?

File: cmd/cloud.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	Example: `
+	k6 cloud script.js`[1:],
+
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		_, _ = BannerColor.Fprint(stdout, Banner+"\n\n")
+		fmt.Fprintf(stdout, "  Deploying script to the cloud..")
```
Comment (robingustafsson) [inline cmd/cloud.go]: Perhaps say "test" here instead of "script"?

File: cmd/cloud.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+		if err := envconfig.Process("k6", &cloudConfig); err != nil {
+			return err
+		}
+
+		token := cloudConfig.Token
+		host := cloudConfig.Host
+
+		if token == "" {
+			return errors.New("Your cloud token is empty, please authenticate using `k6 login cloud`.")
```
Comment (robingustafsson) [inline cmd/cloud.go]: I think we should refer to the token as "Load Impact API token" rather than "cloud token" here.

**Comment by ppcano** [comment]
@liclac what do you think about Robin text changes? They are related to the #390 review.

Please, let's move forward to get this PR merged, I need this for the cloud execution.

File: cmd/cloud.go (hunk 1 above)
Comment (liclac) [reply cmd/cloud.go]: Too long, jeez.

"Start a load test on loadimpact.com"?

File: cmd/cloud.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+var cloudCmd = &cobra.Command{
+	Use:   "cloud",
+	Short: "Execute a test on the cloud",
+	Long: `Execute a test on the cloud.
+
+                This will execute the test on the Load Impact cloud service. Use "k6 login cloud" to authenticate.`,
```
Comment (liclac) [inline cmd/cloud.go]: This indent doesn't look right.

File: cmd/cloud.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	Example: `
+                k6 cloud script.js`[1:],
```
Comment (liclac) [inline cmd/cloud.go]: This indent doesn't look right either.

File: cmd/cloud.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/spf13/cobra"
+)
+
+var cloudCmd = &cobra.Command{
+	Use:   "cloud",
+	Short: "Execute a test on the cloud",
+	Long: `Execute a test on the cloud.
+
+                This will execute the test on the Load Impact cloud service. Use "k6 login cloud" to authenticate.`,
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	Example: `
+                k6 cloud script.js`[1:],
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go (hunk 3 above)
Comment (liclac) [reply cmd/cloud.go]: I like script tbh.

File: cmd/cloud.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		_, _ = BannerColor.Fprint(stdout, Banner+"\n\n")
+		fmt.Fprintf(stdout, "  Deploying script to the cloud..")
+		ticker := time.NewTicker(time.Millisecond * 500)
```
Comment (liclac) [inline cmd/cloud.go]: Add a comment for this, it's clever but took me a moment to get.

File: cmd/cloud.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+		go func() {
+			for range ticker.C {
+				fmt.Fprintf(stdout, ".")
```
Comment (liclac) [inline cmd/cloud.go]: `fmt.Fprint` is gonna be a better match if there's no formatting going on.

File: cmd/cloud.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+		}
+
+		conf := cliConf.Apply(fileConf).Apply(Config{Options: r.GetOptions()}).Apply(envConf).Apply(cliConf)
+		r.SetOptions(conf.Options)
+
+		arc := r.MakeArchive()
+
+		// Overwrite Cloud config
+		cloudConfig := conf.Collectors.Cloud
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+		fileConf, _, err := readDiskConfig()
+		if err != nil {
+			return err
+		}
+
+		cliConf, err := getConfig(cmd.Flags())
+		if err != nil {
+			return err
+		}
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+		// Options
+		fileConf, _, err := readDiskConfig()
+		if err != nil {
+			return err
+		}
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+		cliConf, err := getConfig(cmd.Flags())
+		if err != nil {
+			return err
+		}
+
+		envConf, err := readEnvConfig()
+		if err != nil {
+			return err
+		}
+
```
Comment (liclac) [inline cmd/cloud.go]: No newline.

File: cmd/cloud.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+		conf := cliConf.Apply(fileConf).Apply(Config{Options: r.GetOptions()}).Apply(envConf).Apply(cliConf)
+		r.SetOptions(conf.Options)
+
+		arc := r.MakeArchive()
```
Comment (liclac) [inline cmd/cloud.go]: Maybe create this closer to where it's used.

File: cmd/cloud.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+		// Overwrite Cloud config
+		cloudConfig := conf.Collectors.Cloud
+
+		if err := envconfig.Process("k6", &cloudConfig); err != nil {
+			return err
+		}
+
+		token := cloudConfig.Token
+		host := cloudConfig.Host
```
Comment (liclac) [inline cmd/cloud.go]: Just use the `cloudConfig` members directly.

File: cmd/cloud.go (hunk 4 above)
Comment (liclac) [reply cmd/cloud.go]: Or even "Not logged in, please use `k6 login cloud`".

File: cmd/cloud.go (hunk 10 above)
Comment (liclac) [reply cmd/cloud.go]: Also: start it when you actually start uploading stuff.

File: cmd/cloud.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+		if token == "" {
+			return errors.New("Your cloud token is empty, please authenticate using `k6 login cloud`.")
+		}
+
+		// Start cloud test run
+		client := cloud.NewClient(token, host, Version)
+
+		err = client.ValidateOptions(arc.Options)
```
Comment (liclac) [inline cmd/cloud.go]: This can be a `if err := …; err != nil` one-liner.

File: stats/cloud/api.go [hunk 19]
```diff
@@ -87,6 +91,51 @@ func (c *Client) PushMetric(referenceID string, samples []*Sample) error {
 	return err
 }
 
+func (c *Client) StartCloudTestRun(name string, arc *lib.Archive) (string, error) {
+	requestUrl := fmt.Sprintf("%s/archive-upload", c.baseURL)
+
+	buf := bytes.NewBuffer(nil)
```
Comment (liclac) [inline stats/cloud/api.go]: If you're not preinitialising the buffer with anything, `var buf bytes.Buffer` is better.

https://godoc.org/bytes#Buffer

File: stats/cloud/api.go [hunk 20]
```diff
@@ -112,3 +161,22 @@ func (c *Client) TestFinished(referenceID string, thresholds ThresholdResult, ta
...
+	}{
+		options,
+	}
+
+	req, err := c.NewRequest("POST", url, data)
+	if err != nil {
+		return err
+	}
+
+	err = c.Do(req, nil)
```
Comment (liclac) [inline stats/cloud/api.go]: Can be simplified to: `return c.Do(req, nil)`. (I thought the linter was supposed to catch this…)

File: stats/cloud/api.go [hunk 21]
```diff
@@ -87,6 +91,51 @@ func (c *Client) PushMetric(referenceID string, samples []*Sample) error {
...
+	err = c.Do(req, &ctrr)
+	if err != nil {
+		return "", err
+	}
+
+	path := "runs"
+	if c.token == "" {
+		path = "anonymous"
+	}
+	testUrl := fmt.Sprintf("https://app.loadimpact.com/k6/%s/%s", path, ctrr.ReferenceID)
```
Comment (liclac) [inline stats/cloud/api.go]: Can be simplified to: `return fmt.Sprintf(…), nil`

**Comment by ppcano** [comment]
@liclac `k6 cloud` requested changes have been added.

File: cmd/cloud.go (hunk 1 above)
Comment (liclac) [reply cmd/cloud.go]: "Execute" is inconsistent wording with `run`.

File: cmd/cloud.go [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+	Short: "Execute a test on the cloud",
+	Long: `Execute a test on the cloud.
+
+This will execute the test on the Load Impact cloud service. Use "k6 login cloud" to authenticate.`,
+	Example: `
+	k6 cloud script.js`[1:],
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		_, _ = BannerColor.Fprint(stdout, Banner+"\n\n")
+		fmt.Fprint(stdout, "  Deploying script to the cloud..")
```
Comment (liclac) [inline cmd/cloud.go]: Hm. I'm iffy about the wording here; "Uploading" maybe? "Starting"?

"Deploying" makes me think of something more heavyweight than this; I "deploy" a website, for example.

File: cmd/cloud.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+			return err
+		}
+		envConf, err := readEnvConfig()
+		if err != nil {
+			return err
+		}
+		conf := cliConf.Apply(fileConf).Apply(Config{Options: r.GetOptions()}).Apply(envConf).Apply(cliConf)
+		r.SetOptions(conf.Options)
+
+		// Overwrite Cloud config
```
Comment (liclac) [inline cmd/cloud.go]: Let's just make this "Cloud config" or something, "overwrite" is a strange word to use here.

File: cmd/cloud.go [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+		cloudConfig := conf.Collectors.Cloud
+		if err := envconfig.Process("k6", &cloudConfig); err != nil {
+			return err
+		}
+		if cloudConfig.Token == "" {
+			return errors.New("Not logged in, please use `k6 login cloud`.")
+		}
+
+		// Start cloud test run
```
Comment (liclac) [inline cmd/cloud.go]: Move these two lines below the ticker.

File: cmd/cloud.go [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		name := filepath.Base(filename)
+		testUrl, err := client.StartCloudTestRun(name, arc)
+		if err != nil {
+			return err
+		}
+		ticker.Stop()
+
+		fmt.Fprintf(stdout, "\n\n  execution: %s\n", ui.ValueColor.Sprint("cloud"))
```
Comment (liclac) [inline cmd/cloud.go]: Let's put the "\n\n" in its own `fmt.Fprint` call. This makes the alignment inconsistent in the source file itself.

File: stats/cloud/api.go [hunk 26]
```diff
@@ -83,8 +87,50 @@ func (c *Client) PushMetric(referenceID string, samples []*Sample) error {
 		return err
 	}
 
-	err = c.Do(req, nil)
-	return err
+	return c.Do(req, nil)
+}
+
+func (c *Client) StartCloudTestRun(name string, arc *lib.Archive) (string, error) {
```
Comment (liclac) [inline stats/cloud/api.go]: On second thought, I don't think this should return a URL; return the CreateTestRunResponse directly and assemble the URL elsewhere (maybe make a helper, `URLForResults(refID string, anon bool) string` or somesuch?).

File: stats/cloud/api.go [hunk 27]
```diff
@@ -83,8 +87,50 @@ func (c *Client) PushMetric(referenceID string, samples []*Sample) error {
...
+
+	req, err := http.NewRequest("POST", requestUrl, &buf)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Content-Type", mp.FormDataContentType())
+
+	ctrr := CreateTestRunResponse{}
+	err = c.Do(req, &ctrr)
```
Comment (liclac) [inline stats/cloud/api.go]: You can make this an `if err := …; err != nil`.

**Comment by ppcano** [comment]
@liclac `k6 cloud` requested changes done.

**Comment by liclac** [comment]
@robingustafsson Just waiting on your review to merge o7

**Pull request merged**